src/hkdfchachapoly.rs

```diff
@@ -0,0 +1,198 @@
+pub const NONCE_SIZE: usize = 24;
+// CHACHA_NONCE_SIZE is the size of the nonce used in Chacha
+pub const CHACHA_NONCE_SIZE: usize = 12;
+// TAG_SIZE is the size added from poly1305
+pub const TAG_SIZE: usize = 16;
+// MAX_PLAINTEXT_SIZE is the max size that can be passed into a single call of Seal
+pub const MAX_PLAINTEXT_SIZE: usize = (1 << 38) - 64;
+// MAX_CIPHERTEXT_SIZE is the max size that can be passed into a single call of Open,
+// this differs from plaintext size due to the tag
+pub const MAX_CIPHERTEXT_SIZE: usize = MAX_PLAINTEXT_SIZE - 16;
+// HKDF_INFO is the parameter used internally for Hkdf's info parameter.
+const HKDF_INFO: &str = "TENDERMINT_SECRET_CONNECTION_FRAME_KEY_DERIVE";
+
+impl Aead for HkdfChaChaPoly {
+    fn name(&self) -> &'static str {
+        "HkdfChaChaPoly"
+    }
+
+    fn set(&mut self, key: [u8; 32]) {
+        self.key = key;
+    }
+
+    fn seal(
+        &self,
+        nonce: &[u8; NONCE_SIZE],
+        authtext: &[u8],
+        plaintext: &[u8],
+        out: &mut [u8],
+    ) -> Result<usize, ()> {
+        out[..plaintext.len()].copy_from_slice(plaintext);
+        // if plaintext.len() > MAX_PLAINTEXT_SIZE {
+        //     return error!("Plaintext is greater than the maximum size")
+        // }
+
+        let (subkey, chacha_nonce) = get_subkey_and_chacha_nonce_from_hkdf(&self.key, &nonce);
+        let sealing_key = aead::SealingKey::new(&aead::CHACHA20_POLY1305, &subkey).unwrap();
+        let res = aead::seal_in_place(
+            &sealing_key,
+            &chacha_nonce,
+            authtext,
+            &mut out[..plaintext.len() + TAG_SIZE],
+            16,
+        );
+        Ok(plaintext.len() + TAG_SIZE)
+    }
+
+    fn open(
+        &self,
+        nonce: &[u8; NONCE_SIZE],
+        authtext: &[u8],
+        ciphertext: &[u8],
+        out: &mut [u8],
+    ) -> Result<usize, ()> {
+        // if ciphertext.len() > MAX_CIPHERTEXT_SIZE {
+        //     return error!("Plaintext is greater than the maximum size")
+        // }
+
+        let (subkey, chacha_nonce) = get_subkey_and_chacha_nonce_from_hkdf(&self.key, &nonce);
+        let opening_key = aead::OpeningKey::new(&aead::CHACHA20_POLY1305, &subkey).unwrap();
+        // optimize if the provided buffer is sufficiently large
+        if out.len() >= ciphertext.len() {
+            let in_out = &mut out[..ciphertext.len()];
+            in_out.copy_from_slice(ciphertext);
+
+            let len = aead::open_in_place(&opening_key, &chacha_nonce, authtext, 0, in_out)
+                .map_err(|_| ())?
+                .len();
+
+            Ok(len)
+        } else {
+            let mut in_out = ciphertext.to_vec();
+
+            let out0 = aead::open_in_place(&opening_key, &chacha_nonce, authtext, 0, &mut in_out)
+                .map_err(|_| ())?;
+            out[..out0.len()].copy_from_slice(out0);
+            Ok(out0.len())
+        }
+    }
+}
+
+pub fn new_hkdfchachapoly(key: [u8; 32]) -> HkdfChaChaPoly {
```

Comment:
make this a `new` method inside `impl aead`

Reply:
Should I make aead implement sized then also, so that this works? 
Otherwise, I can do the same as mcginty's snow, which make aead implement default, and uses a set command.